### PR TITLE
Use MatrixView and Span in KinDynComputations class

### LINKED ISSUE #687: Improve the use of `const` in KinDynComputation
I noticed that some functions in `KinDynComputation` are not const even if they don't modify any member of the class. For instance the function [`bool iDynTree::KinDynComputations::getJointPos()`](https://robotology.github.io/idyntree/devel/classiDynTree_1_1KinDynComputations.html#a3154c238fcccc126e56d95e40e2fd832) is not considered as a `const` method.

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
This PR implements several functions to allow using `MatrixView` and `span` object into `KinDynComputations` object.

The original functions remain, however, this PR brakes some ABI because of #687. This would not be a huge problem since the original test passes without any changes (@traversaro correct me if I'm mistaken). 

I've also implemented a new test to check the possibility to call `KinDynComputations` function with `non-iDynTree` objects. 

This PR closes #687 #716

### CODE REVIEW

**Comment by GiulioRomualdi** [comment]
CI is failing while generating the Python bindings

**Comment by traversaro** [comment]
> The original functions remain, however, this PR brakes some ABI because of #687. This would not be a huge problem since the original test passes without any changes (@traversaro correct me if I'm mistaken).

Breaking the ABI is not a problem, as in any case the `devel` branch will be release as a new major version.

**Comment by traversaro** [comment]
> 
> 
> CI is failing while generating the Python bindings

I think the trick of declaring the method arguments as `iDynTree::MatrixView` instead of `MatrixView` (consistently with the rest of the KinDynComputations methods arguments) could work here.

File: src/core/include/iDynTree/Core/VectorDynSize.h [hunk 1]
```diff
@@ -80,6 +80,18 @@ namespace iDynTree
          */
         VectorDynSize(const VectorDynSize& vec);
 
+#if !defined(SWIG_VERSION) || SWIG_VERSION >= 0x030000
```
Comment (traversaro) [inline src/core/include/iDynTree/Core/VectorDynSize.h:83]: Just to understand, why the `SWIG_VERSION >= 0x030000` ?

**Comment by traversaro** [comment]
I still need to check all the details, but this PR is **majestic**!

File: src/core/include/iDynTree/Core/VectorDynSize.h (hunk 1 above)
Comment (GiulioRomualdi) [reply src/core/include/iDynTree/Core/VectorDynSize.h:83]: Actually I don't know. I'm not an expert of `SWIG` so I copied  what was done here:
https://github.com/robotology/idyntree/blob/c0afdd969bf5e7629b4c39df54880e0785fc7753/src/core/include/iDynTree/Core/VectorDynSize.h#L113-L124

If I've to tell the truth I don't know why we need this. Probably @diegoferigo can give me an explanation

Comment (traversaro) [reply src/core/include/iDynTree/Core/VectorDynSize.h:83]: I blamed the code and it seems that those if come from https://github.com/robotology/idyntree/pull/476 . I am not sure if anyone is still testing or interested at all in SWIG 2.x compatibility, but for sure they are not harming anyone.

Comment (diegoferigo) [reply src/core/include/iDynTree/Core/VectorDynSize.h:83]: I think all the systems we support have at least swig 3. I'd be in favour of removing these preprocessor statements and bump the required swig version in CMake.

File: src/high-level/include/iDynTree/KinDynComputations.h [hunk 2]
```diff
@@ -256,6 +257,25 @@ class KinDynComputations {
                                             const iDynTree::FrameIndex frameIndex,
                                             iDynTree::MatrixDynSize & outJacobianPattern) const;
 
+    /**
+     * Returns the sparsity pattern of the relative Jacobian for the specified frames
+     * (MatrixView implementation)
+     *
+     * The resulting matrix must have the same size of the free floating Jacobian (6 x #DoFs)
```
Comment (traversaro) [inline src/high-level/include/iDynTree/KinDynComputations.h:264]:
```suggestion
     * The resulting matrix must have the same size of the relative Jacobian (6 x #DoFs)
```

File: src/core/include/iDynTree/Core/VectorDynSize.h (hunk 1 above)
Comment (GiulioRomualdi) [reply src/core/include/iDynTree/Core/VectorDynSize.h:83]: > I think all the systems we support have at least swig 3. I'd be in favour of removing these preprocessor statements and bump the required swig version in CMake.

So if you (@traversaro) agree I can add this modification. Let me know if it's better to open another PR (probably yes)

**Comment by S-Dafarra** [comment]
Just a nitpick. I have noticed that in the interfaces you passed the MatrixView and the Span mostly by const reference. Would not make sense to pass by copy instead, as we usually do with pointers?

**Comment by GiulioRomualdi** [comment]
> Just a nitpick. I have noticed that in the interfaces you passed the MatrixView and the Span mostly by const reference. Would not make sense to pass by copy instead, as we usually do with pointers?

I did this because it may happen that the user has already a `Span` or a `MatrixView` object (i.e. the object is not created in place) and in this case, we can avoid creating a new object every time. 

For instance
https://github.com/robotology/idyntree/blob/5cc8c2ff4f8db4c8349f7135f8dbdaad2a9d32ec/src/model/src/Jacobians.cpp#L23-L30

is called in 

https://github.com/robotology/idyntree/blob/c0afdd969bf5e7629b4c39df54880e0785fc7753/src/high-level/src/KinDynComputations.cpp#L1800-L1803

Where `outJacobian` is already an `iDynTree::MatrixView<double>`

**Comment by S-Dafarra** [comment]
> > Just a nitpick. I have noticed that in the interfaces you passed the MatrixView and the Span mostly by const reference. Would not make sense to pass by copy instead, as we usually do with pointers?
> 
> I did this because it may happen that the user has already a `Span` or a `MatrixView` object (i.e. the object is not created in place) and in this case, we can avoid creating a new object every time.
> 
> For instance
> https://github.com/robotology/idyntree/blob/5cc8c2ff4f8db4c8349f7135f8dbdaad2a9d32ec/src/model/src/Jacobians.cpp#L23-L30
> 
> is called in
> 
> https://github.com/robotology/idyntree/blob/c0afdd969bf5e7629b4c39df54880e0785fc7753/src/high-level/src/KinDynComputations.cpp#L1800-L1803
> 
> Where `outJacobian` is already an `iDynTree::MatrixView<double>`

I see, but the copy of these objects is basically inexpensive, just a few bytes. Isn't the const reference preventing to use temporary Span or MatrixView objects?

File: src/core/include/iDynTree/Core/VectorDynSize.h (hunk 1 above)
Comment (GiulioRomualdi) [reply src/core/include/iDynTree/Core/VectorDynSize.h:83]: In `Span.h` there are a lot of `#ifndef SWIG`, e.g. 
https://github.com/robotology/idyntree/blob/5b2b487b123155dfadccef92358cccfea10e3ef2/src/core/include/iDynTree/Core/Span.h#L362-L371

I don't know if this may cause a problems. As far as I understood if `SWIG` is defined that part of the code is not defined, consequentially we cannot use that constructor in the code. Am I missing something?

**Comment by GiulioRomualdi** [comment]
@S-Dafarra 
> Isn't the const reference preventing to use temporary Span or MatrixView objects

In theory no, this code 
```cpp
void foo(const iDynTree::Span<double> & span)
{
   ...
}

...
iDynTree::VectorDynSize v(10);
foo(make_span(v.data(), v.size()):
```
should copile.

See here for example
https://github.com/robotology/idyntree/blob/c0afdd969bf5e7629b4c39df54880e0785fc7753/src/high-level/tests/KinDynComputationsMatrixViewAndSpanUnitTest.cpp#L122

where `getModelVel()` is defined as:
https://github.com/robotology/idyntree/blob/c0afdd969bf5e7629b4c39df54880e0785fc7753/src/high-level/include/iDynTree/KinDynComputations.h#L463

**Comment by S-Dafarra** [comment]
I guess I don't like ``const`` objects that are outputs, but on the other hand, this prevents accidental modifications on the ``Span`` itself, avoiding stupid bugs when using the copy operator expecting the content to be copied. So, I guess it is ok as it is.

File: src/core/include/iDynTree/Core/VectorDynSize.h (hunk 1 above)
Comment (traversaro) [reply src/core/include/iDynTree/Core/VectorDynSize.h:83]: > I think all the systems we support have at least swig 3. I'd be in favour of removing these preprocessor statements and bump the required swig version in CMake.

Also in 2018 that was the case, but some old users asked me to support SWIG 2 for their particular workflow. I think that also them do not use SWIG 2 anymore, but let's wait for them to appear and confirm they do not use SWIG 2 anymore. 

Completely unrelated mention: @francesco-romano . :D

Comment (traversaro) [reply src/core/include/iDynTree/Core/VectorDynSize.h:83]: > In Span.h there are a lot of #ifndef SWIG, e.g.

Those are different, as they are active for all possible version of SWIG. 

> As far as I understood if SWIG is defined that part of the code is not defined, consequentially we cannot use that constructor in the code. Am I missing something?

The `SWIG` preprocessor define is only defined by the SWIG compiler used to generate the bindings, so it is perfectly fine to use those constructor in the normal code, as `SWIG` is never defined in the regular compilation calls, even if SWIG generation is enabled at the CMake level.

Comment (francesco-romano) [reply src/core/include/iDynTree/Core/VectorDynSize.h:83]: If you can keep the `ifdef` on the SWIG version it would be great.
We currently still use SWIG 2.0

**Comment by francesco-romano** [comment]
Regarding by-value vs by-reference discussion, in `absl` we usually pass by value. Probably the same convention is for `std::span`

File: src/core/src/MatrixDynSize.cpp [hunk 3]
```diff
@@ -42,6 +42,29 @@ MatrixDynSize::MatrixDynSize(unsigned int _rows,
     zero();
 }
 
+MatrixDynSize::MatrixDynSize(const MatrixView<const double>& other) : m_rows(other.rows()),
```
Comment (francesco-romano) [inline src/core/src/MatrixDynSize.cpp:45]: I am a bit rusty in iDynTree.. is the ordering defined only for the sparse matrix?
Are dense matrices always row major?

File: src/high-level/include/iDynTree/KinDynComputations.h [hunk 4]
```diff
@@ -256,6 +257,25 @@ class KinDynComputations {
...
+     * @param refFrameIndex refence frame for the Jacobian
+     * @param frameIndex Jacobian frame
+     * @param outJacobianPattern a MatrixView containing an already existing memory.
+     * @warning outJacobianPattern size must be 6 x #DoFs, it is not possible to resize a
+     * MatrixView element.
+     * @return true on success. False otherwise
+     */
+    bool getRelativeJacobianSparsityPattern(const iDynTree::FrameIndex refFrameIndex,
+                                            const iDynTree::FrameIndex frameIndex,
+                                            const MatrixView<double> & outJacobianPattern) const;
```
Comment (francesco-romano) [inline src/high-level/include/iDynTree/KinDynComputations.h:277]: Not sure about having a `const` on an output parameter.

File: src/high-level/include/iDynTree/KinDynComputations.h [hunk 5]
```diff
@@ -271,6 +291,23 @@ class KinDynComputations {
...
+     * - 0: the element will always have 0 (for every robot configuration)
+     * - 1: it exists a robot configuration such that the element have a value different from zero.
+     * @param frameIndex Jacobian frame
+     * @param outJacobianPattern the Jacobian sparsity pattern
+     * @warning outJacobianPattern size must be 6 x (#DoFs + 6), it is not possible to resize a
+     * MatrixView element.
+     * @return true on success. False otherwise
+     */
+    bool getFrameFreeFloatingJacobianSparsityPattern(const FrameIndex frameIndex,
+                                                     const MatrixView<double> & outJacobianPattern) const;
```
Comment (francesco-romano) [inline src/high-level/include/iDynTree/KinDynComputations.h:309]: Same comment as before

File: src/high-level/include/iDynTree/KinDynComputations.h [hunk 6]
```diff
@@ -328,20 +408,60 @@ class KinDynComputations {
...
+    /**
+     * Get the n joint position of the model.
+     * @warning the Span object should point an already existing memory. Memory allocation and resizing cannot be achieved with this kind of objects.
+     * @return true if all went well, false otherwise.
+     */
+    bool getJointPos(const iDynTree::Span<double> &q) const;
 
-    bool getJointPos(iDynTree::VectorDynSize &q);
-    bool getJointVel(iDynTree::VectorDynSize &dq);
+    bool getJointVel(iDynTree::VectorDynSize &dq) const;
```
Comment (francesco-romano) [inline src/high-level/include/iDynTree/KinDynComputations.h:442]: If:
- performance tests do not show regression
- you do not have a particular reason to use this,

I suggest to deprecate the explicit vector version and keep only methods accepting `Span` or `MatrixView`: cleaner interface (of course, where it is possible, i.e. not in `getWorldTransform`)

File: src/high-level/include/iDynTree/KinDynComputations.h [hunk 7]
```diff
@@ -371,13 +491,37 @@ class KinDynComputations {
...
+    iDynTree::Transform getWorldTransform(const std::string & frameName);
+
+    /**
+     * Version of getWorldTransform where the frame is specified by name
+     * (MatrixView and Span implementation).
+     *
+     * @note For real time code use the method that takes an integer.
+     *
+     */
+    bool getWorldTransform(const std::string & frameName,
```
Comment (francesco-romano) [inline src/high-level/include/iDynTree/KinDynComputations.h:522]: Maybe: do you need to access spans of `Transform` objects?
If you have an immediate need for this (meaning: you have some code that MUST use the span here), then, forget about this comment.
Otherwise, I suggest to drop the span version of methods accepting/returning Transform objects. 
These objects are not directly mapped to matrices, and the resulting API  is ugly and confusing

File: src/core/include/iDynTree/Core/VectorDynSize.h (hunk 1 above)
Comment (traversaro) [reply src/core/include/iDynTree/Core/VectorDynSize.h:83]: Thanks @francesco-romano !

**Comment by traversaro** [comment]
> 
> 
> Regarding by-value vs by-reference discussion, in `absl` we usually pass by value. Probably the same convention is for `std::span`

Yes, sometimes the by-value for small object can be more cache-friendly, however both version are ok for me.

**Review by francesco-romano — CHANGES_REQUESTED** [review]
Regarding by-value vs by-reference I could not find anything on `std::span`, but this might be useful
https://abseil.io/tips/93

> As with absl::string_view, it is usually better to pass absl::Span by value when used as a function parameter; this form is slightly faster than passing by const reference (on most platforms), and produces smaller code

File: src/core/src/MatrixDynSize.cpp (hunk 3 above)
Comment (traversaro) [reply src/core/src/MatrixDynSize.cpp:45]: Yes, dense matrices are always row major.

File: src/high-level/include/iDynTree/KinDynComputations.h (hunk 6 above)
Comment (traversaro) [reply src/high-level/include/iDynTree/KinDynComputations.h:442]: The explicit vector version has always had a slightly different semantics: it resize the vector to be of the correct size.

File: src/high-level/include/iDynTree/KinDynComputations.h (hunk 7 above)
Comment (traversaro) [reply src/high-level/include/iDynTree/KinDynComputations.h:522]: The idea is to be able to use Eigen-matrix only (or any other matrix library) to interface with KinDynComputations. In that case, you need in any case to transform for your specific representation of Transform (in our codebases almost always 4x4 homegenous matrix) having a span variant of the Transform methods permits to reduce to the minimum the boilerplate necessary to use iDynTree classes. If you think about it, it is a bit similar to what we do in WB-Toolbox, where when we need to have a Simulink signal that represent an iDynTree::Transform, we use a 4x4 matrix.

File: src/high-level/include/iDynTree/KinDynComputations.h (hunk 6 above)
Comment (francesco-romano) [reply src/high-level/include/iDynTree/KinDynComputations.h:442]: I am fine in keeping things as they are in this PR, but I strongly suggest revisiting this later (but not too late). 
API bloating is a problem. 
Deprecating and suggesting the user to pass a pre-allocated span should not be a problem. 
Open an issue to track this.

File: src/high-level/include/iDynTree/KinDynComputations.h (hunk 7 above)
Comment (francesco-romano) [reply src/high-level/include/iDynTree/KinDynComputations.h:522]: Fine. 
But as you probably know (as you wrote the `Transform` class), the 4x4 is just one of the possible representations. And the class is kind of clear on this. 
Adding a span version adds confusion IMHO. 
If this is needed I suggest to change the name of the method at least, to make this explicit. 
For example `getWorldTransformComponents`

Comment (traversaro) [reply src/high-level/include/iDynTree/KinDynComputations.h:522]: Ahh, I see the main point here. I think on the other hand that having exactly the same method names is useful for migration, but indeed having the same name but a different list of arguments is indeed confusing. Could it make sense to have for methods with the same names the same list of arguments, so having in place of a `iDynTree::Transform`  a 4x4 MatrixView that represent the corresponding homogenous transform? Probably this applies to other parts of the API where we have a `iDynTree::Transform`, for example `setRobotState` .

File: src/high-level/include/iDynTree/KinDynComputations.h (hunk 6 above)
Comment (traversaro) [reply src/high-level/include/iDynTree/KinDynComputations.h:442]: >  Deprecating and suggesting the user to pass a pre-allocated span should not be a problem.

This API is quite new, and in general we do not know if the use of Span/MatrixView is going to create unforeseeable problems. For example, I don't know if the implicit conversion from VectorDynSize to Span works correctly in the Matlab/Python bindings as it works in C++. So I opened an issue to track this in https://github.com/robotology/idyntree/issues/737, but I personally would like to try this API extensively in the wild before deprecating and removing the old methods.

**Comment by diegoferigo** [comment]
>> CI is failing while generating the Python bindings
>
>I think the trick of declaring the method arguments as `iDynTree::MatrixView` instead of `MatrixView` (consistently with the rest >of the KinDynComputations methods arguments) could work here.

Yep, this could be enough. Let's remember to also instantiate the `MatrixView` template as done for `Span` in the `.i` files (note: I never tried to use Span from Python):

https://github.com/robotology/idyntree/blob/a04fe2b47e1f846e996c1a64cf6b7a3060baeb4e/bindings/iDynTree.i#L222-L224

File: src/high-level/include/iDynTree/KinDynComputations.h (hunk 6 above)
Comment (diegoferigo) [reply src/high-level/include/iDynTree/KinDynComputations.h:442]: @traversaro I read this last comment after I wrote https://github.com/robotology/idyntree/issues/737#issuecomment-691930285, I think that what I wrotefor Python also applied to Matlab, good point. We have a lot of code that use the Matlab bindings that I'm not sure we want to break. Python situation is instead different.

**Comment by diegoferigo** [comment]
From [the logs](https://github.com/robotology/idyntree/runs/1112145435#step:8:7061), it seems that swig does not like the following:

https://github.com/robotology/idyntree/blob/a04fe2b47e1f846e996c1a64cf6b7a3060baeb4e/src/core/include/iDynTree/Core/MatrixView.h#L36

I foresee a struggle with SFINAE :/

**Comment by traversaro** [comment]
If necessary, you can skip any part of the code in the SWIG preprocessor with the `ifndef SWIG` instruction.

File: src/high-level/include/iDynTree/KinDynComputations.h (hunk 7 above)
Comment (francesco-romano) [reply src/high-level/include/iDynTree/KinDynComputations.h:522]: This is not a trivial decision.
What I mean is:
- I consider it a safe change (modulo your comments on my other comment) to change a parameter from a vector/array to a span-like object. Conversion should be either implicit or simple by helper functions. In this case I do not see a reason to change name (nor to keep the old functions around).
- This is a different case. You can have a version taking a 4x4 view, maybe extend `Transform` with a constructor taking a 4x4 matrix, but for users who wants to use the explicit `Transform` (and I see the value in it) it will be a 2-steps procedure, plus waisting an intermediary object. This is why I was suggesting making it clear that they are two different methods

**Comment by GiulioRomualdi** [comment]
In theory, https://github.com/robotology/idyntree/pull/736/commits/481fb572160f76c05e0646af49b36feb6e9b9c8f should fix the python bindings generation. Let me know if I'm missing something @diegoferigo @traversaro

File: src/high-level/include/iDynTree/KinDynComputations.h (hunk 7 above)
Comment (traversaro) [reply src/high-level/include/iDynTree/KinDynComputations.h:515]: >  You can have a version taking a 4x4 view, maybe extend Transform with a constructor taking a 4x4 matrix, but for users who wants to use the explicit Transform (and I see the value in it) it will be a 2-steps procedure, plus waisting an intermediary object. This is why I was suggesting making it clear that they are two different methods

At the moment, people will continue to use the Transform variant as usual without any kind of 2-steps procedure, as we are going to leave both variant around. If we deprecate/remove the Transform version and just leave the span version (that now to be honest seems unlikely to me) we will address about this aspects. 

In a nutshell, I think your comment would make sense if we were going to remove the non-span/view versions soon, but as mention in the other thread for now we will keep both versions to see how hergonomic the use of the span/view version is.

Comment (francesco-romano) [reply src/high-level/include/iDynTree/KinDynComputations.h:515]: No I am not suggesting to remove the `Transform` version (while I was suggesting to remove the `Vector` version to the other methods).
If you change the span version to get a span to a 4x4 matrix, (and document that that is a homogenous transformation matrix) then it looks good to me :)

**Comment by GiulioRomualdi** [comment]
This is a recap of the status of this PR.
1. Following @S-Dafarra and @francesco-romano suggestions the `MatrixView` and `Span` objects are now passed by copy 
2. The transform is now stored in a 4x4 `MatrixView` instead of 3-d `span` + 3x3 `MatrixView`.  
3. The python bindings are correctly generated. 
4. The PR is working on Linux macOS and Windows

Waiting for further requests/suggestions 
@traversaro @S-Dafarra @prashanthr05

File: src/high-level/src/KinDynComputations.cpp [hunk 8]
```diff
@@ -856,21 +1096,61 @@ Twist KinDynComputations::getBaseTwist()
...
+
+bool KinDynComputations::getJointVel(VectorDynSize& dq) const
 {
     dq.resize(pimpl->m_robot_model.getNrOfDOFs());
     dq = this->pimpl->m_vel.jointVel();
     return true;
 }
 
-bool KinDynComputations::getModelVel(VectorDynSize& nu)
+bool KinDynComputations::getJointVel(Span<double>dq) const
```
Comment (S-Dafarra) [inline src/high-level/src/KinDynComputations.cpp:1140]:
```suggestion
bool KinDynComputations::getJointVel(Span<double> dq) const
```

File: src/high-level/src/KinDynComputations.cpp [hunk 9]
```diff
@@ -856,21 +1096,61 @@ Twist KinDynComputations::getBaseTwist()
...
+
+bool KinDynComputations::getJointPos(VectorDynSize& q) const
 {
     q.resize(this->pimpl->m_robot_model.getNrOfPosCoords());
     toEigen(q) = toEigen(this->pimpl->m_pos.jointPos());
     return true;
 }
 
-bool KinDynComputations::getJointVel(VectorDynSize& dq)
+bool KinDynComputations::getJointPos(Span<double>q) const
```
Comment (S-Dafarra) [inline src/high-level/src/KinDynComputations.cpp:1120]:
```suggestion
bool KinDynComputations::getJointPos(Span<double> q) const
```

File: src/high-level/include/iDynTree/KinDynComputations.h [hunk 10]
```diff
@@ -328,20 +405,58 @@ class KinDynComputations {
...
+    bool getBaseTwist(iDynTree::Span<double> base_velocity) const;
+
+    bool getJointPos(iDynTree::VectorDynSize &q) const;
+
+    /**
+     * Get the n joint position of the model.
+     * @warning the Span object should point an already existing memory. Memory allocation and resizing cannot be achieved with this kind of objects.
+     * @return true if all went well, false otherwise.
+     */
+    bool getJointPos(iDynTree::Span<double>q) const;
```
Comment (S-Dafarra) [inline src/high-level/include/iDynTree/KinDynComputations.h:435]:
```suggestion
    bool getJointPos(iDynTree::Span<double> q) const;
```

**Review by traversaro — APPROVED** [review]

File: src/high-level/include/iDynTree/KinDynComputations.h [hunk 11]
```diff
@@ -328,20 +405,58 @@ class KinDynComputations {
...
+
+    bool getJointVel(iDynTree::VectorDynSize &dq) const;
+
+    /**
+     * Get the n joint velocity of the model.
+     * @warning the Span object should point an already existing memory. Memory allocation and resizing cannot be achieved with this kind of objects.
+     * @return true if all went well, false otherwise.
+     */
+    bool getJointVel(iDynTree::Span<double>dq) const;
```
Comment (S-Dafarra) [inline src/high-level/include/iDynTree/KinDynComputations.h:444]:
```suggestion
    bool getJointVel(iDynTree::Span<double> dq) const;
```

**Comment by traversaro** [comment]
For me the PR is ok! Can you add some details on this change in the CHANGELOG ?

**Review by S-Dafarra — COMMENTED** [review]
I have also seen that somewhere the const reference version is used in the constructors of objects like ``MatrixDynSize``, ``VectorDynSize``, ..., and in the assignment operators. Up yo you if you want to change it also there.

**Review by diegoferigo — APPROVED** [review]

**Comment by GiulioRomualdi** [comment]
@S-Dafarra 
> I have also seen that somewhere the const reference version is used in the constructors of objects like MatrixDynSize, VectorDynSize, ..., and in the assignment operators. Up yo you if you want to change it also there.

Ops I completely forgot them, working on it

**Comment by GiulioRomualdi** [comment]
> @S-Dafarra
> 
> > I have also seen that somewhere the const reference version is used in the constructors of objects like MatrixDynSize, VectorDynSize, ..., and in the assignment operators. Up yo you if you want to change it also there.
> 
> Ops I completely forgot them, working on it

Done 1064a27

**Comment by GiulioRomualdi** [comment]
@traversaro feel free to merge with squash

File: src/high-level/include/iDynTree/KinDynComputations.h [hunk 12]
```diff
@@ -328,20 +405,58 @@ class KinDynComputations {
...
 
     /**
      * Get the n+6 velocity of the model.
-     * Obtained by stacking the output of getBaseTwist and of getJointVel .
+     * Obtained by stacking the output of getBaseTwist and of getJointVel.
+     * @warning the Span object should point an already existing memory. Memory allocation and resizing cannot be achieved with this kind of objects.
+     * @return true if all went well, false otherwise.
      */
-    bool getModelVel(iDynTree::VectorDynSize &nu);
+    bool getModelVel(iDynTree::Span<double>nu) const;
```
Comment (S-Dafarra) [inline src/high-level/include/iDynTree/KinDynComputations.h:458]:
```suggestion
    bool getModelVel(iDynTree::Span<double> nu) const;
```

**Review by S-Dafarra — APPROVED** [review]

File: src/high-level/include/iDynTree/KinDynComputations.h [hunk 13]
```diff
@@ -1,4 +1,4 @@
-/*
+ /*
```
Comment (francesco-romano) [inline src/high-level/include/iDynTree/KinDynComputations.h:1]: Remove this whitespace

File: src/high-level/include/iDynTree/KinDynComputations.h [hunk 14]
```diff
@@ -299,14 +346,32 @@ class KinDynComputations {
      * @param base_velocity The twist (linear/angular velocity) of the base, expressed with the convention specified by the used FrameVelocityConvention.
      * @param s_dot a vector of getNrOfDegreesOfFreedom() joint velocities (in rad/sec)
      * @param world_gravity a 3d vector of the gravity acceleration vector, expressed in the world/inertial frame.
-     *
+     * @return true if all went well, false otherwise.
```
Comment (francesco-romano) [inline src/high-level/include/iDynTree/KinDynComputations.h:349]: on success ?

File: src/high-level/include/iDynTree/KinDynComputations.h [hunk 15]
```diff
@@ -299,14 +346,32 @@ class KinDynComputations {
...
+     * Set the state for the robot (floating base) (Span and MatrixView implementation)
+     *
+     * @param world_T_base the homogeneous transformation that transforms position vectors expressed in the base reference frame
+     *                      in position frames expressed in the world reference frame (i.e. pos_world = world_T_base*pos_base).
+     * @param s a vector of getNrOfDegreesOfFreedom() joint positions (in rad)
+     * @param base_velocity The twist (linear/angular velocity) of the base, expressed with the convention specified by the used FrameVelocityConvention.
+     * @param s_dot a vector of getNrOfDegreesOfFreedom() joint velocities (in rad/sec)
+     * @param world_gravity a 3d vector of the gravity acceleration vector, expressed in the world/inertial frame.
+     * @warning the Span and the MatrixView objects should point an already existing memory. Memory allocation and resizing cannot be achieved with this kind of objects.
+     * @return true if all went well, false otherwise.
```
Comment (francesco-romano) [inline src/high-level/include/iDynTree/KinDynComputations.h:367]: on success?

File: src/high-level/include/iDynTree/KinDynComputations.h [hunk 16]
```diff
@@ -318,6 +383,18 @@ class KinDynComputations {
                        const iDynTree::VectorDynSize &s_dot,
                        const iDynTree::Vector3& world_gravity);
 
+    /**
+     * Set the state for the robot (fixed base)
+     * Same as setRobotState, but with:
+     *  world_T_base      = iDynTree::Transform::Identity()
+     *  base_velocity     = iDynTree::Twist::Zero();
+     * @warning the Span objects should point an already existing memory. Memory allocation and resizing cannot be achieved with this kind of objects.
+     * @return true if all went well, false otherwise.
```
Comment (francesco-romano) [inline src/high-level/include/iDynTree/KinDynComputations.h:392]: on success?

File: src/high-level/include/iDynTree/KinDynComputations.h [hunk 17]
```diff
@@ -371,13 +486,34 @@ class KinDynComputations {
      */
     iDynTree::Transform getWorldTransform(const iDynTree::FrameIndex frameIndex);
 
+    /**
+     * Return the transform where the frame is the frame
+     * specified by frameIndex, and the reference frame is the world one
+     * (world_T_frame).
+     * @param world_T_frame
+     * @warning the Span and the MatrixView objects should point an already existing memory. Memory allocation and resizing cannot be achieved with this kind of objects.
+     * @return true if all went well, false otherwise.
```
Comment (francesco-romano) [inline src/high-level/include/iDynTree/KinDynComputations.h:495]: also here. Please check all the doc

File: src/high-level/include/iDynTree/KinDynComputations.h [hunk 18]
```diff
@@ -371,13 +486,34 @@ class KinDynComputations {
...
      *
      * @note For real time code use the method that takes an integer.
      *
      */
-    iDynTree::Transform getWorldTransform(std::string frameName);
+    iDynTree::Transform getWorldTransform(const std::string & frameName);
+
+    /**
+     * Version of getWorldTransform where the frame is specified by name
+     * (MatrixView and Span implementation).
```
Comment (francesco-romano) [inline src/high-level/include/iDynTree/KinDynComputations.h:510]: This method takes a MatrixView, not a Span, so drop that from the comment.
Also, specify that the span should point to a 4x4 matrix in row-major format representing a homogenous transformation

File: src/high-level/include/iDynTree/KinDynComputations.h [hunk 19]
```diff
@@ -396,6 +532,17 @@ class KinDynComputations {
     iDynTree::Transform getRelativeTransform(const iDynTree::FrameIndex refFrameIndex,
                                              const iDynTree::FrameIndex frameIndex);
 
+    /**
+     * Return the transform where the frame is the frame
+     * specified by frameIndex, and the reference frame is the one specified
+     * by refFrameIndex (refFrame_H_frame).
+     * @warning the Span and the MatrixView objects should point an already existing memory. Memory allocation and resizing cannot be achieved with this kind of objects.
```
Comment (francesco-romano) [inline src/high-level/include/iDynTree/KinDynComputations.h:539]: See the comment in the `getWorldTransform` method.
Also, be consistent with the documentation. 
I _think_ it is known that a span cannot be resized, but if you think the user should know this, add it as class comment as you have lots of methods taking Spans/MatrixViews

File: src/high-level/include/iDynTree/KinDynComputations.h [hunk 20]
```diff
@@ -458,12 +666,53 @@ class KinDynComputations {
...
     bool getFrameFreeFloatingJacobian(const FrameIndex frameIndex,
                                       iDynTree::MatrixDynSize & outJacobian);
 
+    /**
+     * Compute the free floating jacobian for a given frame for the given representaiton (MatrixView implementation).
+     *
+     * @warning the MatrixView objects should point an already existing memory. Memory allocation and resizing cannot be achieved with this kind of objects.
+     * @return true if all went well, false otherwise.
+     */
+    bool getFrameFreeFloatingJacobian(const std::string & frameName,
```
Comment (francesco-romano) [inline src/high-level/include/iDynTree/KinDynComputations.h:705]: nitpick: you probably do not have to replicate both versions (with `std::string` and `FrameIndex`)?

**Review by francesco-romano — CHANGES_REQUESTED** [review]

File: src/high-level/include/iDynTree/KinDynComputations.h (hunk 18 above)
Comment (traversaro) [reply src/high-level/include/iDynTree/KinDynComputations.h:510]: >  Also, specify that the span should point to a 4x4 matrix in ~~~row-major format~~~ representing a homogenous transformation

👍  for adding the details, however  `MatrixView` support both `row-major` and `column-major`, so there is no need to specify that!

File: src/high-level/include/iDynTree/KinDynComputations.h (hunk 20 above)
Comment (traversaro) [reply src/high-level/include/iDynTree/KinDynComputations.h:705]: We wanted to have exact corresponding method between Span and non-Span, to simplify migration to the Span version.

File: src/high-level/include/iDynTree/KinDynComputations.h (hunk 16 above)
Comment (GiulioRomualdi) [reply src/high-level/include/iDynTree/KinDynComputations.h:392]: Here I tried to follow the already existing documentation. 
E.g. 
https://github.com/robotology/idyntree/blob/0d71b87ffbdd2953d9df34d0f61edf24c73d2ad5/src/high-level/include/iDynTree/KinDynComputations.h#L218

https://github.com/robotology/idyntree/blob/0d71b87ffbdd2953d9df34d0f61edf24c73d2ad5/src/high-level/include/iDynTree/KinDynComputations.h#L289

https://github.com/robotology/idyntree/blob/0d71b87ffbdd2953d9df34d0f61edf24c73d2ad5/src/high-level/include/iDynTree/KinDynComputations.h#L690

File: src/high-level/include/iDynTree/KinDynComputations.h (hunk 17 above)
Comment (GiulioRomualdi) [reply src/high-level/include/iDynTree/KinDynComputations.h:495]: See https://github.com/robotology/idyntree/pull/736#discussion_r488465087

File: src/high-level/include/iDynTree/KinDynComputations.h (hunk 16 above)
Comment (GiulioRomualdi) [reply src/high-level/include/iDynTree/KinDynComputations.h:392]: @traversaro let me know if I should change the entire documentation as:
```
@return true in case of success, false otherwise.
```

Comment (traversaro) [reply src/high-level/include/iDynTree/KinDynComputations.h:392]: I agree in general that `in case of success` is cleaner, but I think we can do this change in a different PR. Here adding the message consistently with the rest of the class make sense.

Comment (GiulioRomualdi) [reply src/high-level/include/iDynTree/KinDynComputations.h:392]: Ok we can open a new PR once this has been merged

Comment (francesco-romano) [reply src/high-level/include/iDynTree/KinDynComputations.h:392]: SG.

**Review by francesco-romano — APPROVED** [review]

**Comment by traversaro** [comment]
Thanks to everyone!
